// standardized-audio-context/build/es2015/factories/media-stream-audio-source-node-constructor.js
import { getNativeContext } from '../helpers/get-native-context';
const DEFAULT_OPTIONS = {
    channelCount: 2,
    channelCountMode: 'max',
    channelInterpretation: 'speakers'
};
export const createMediaStreamAudioSourceNodeConstructor = (noneAudioDestinationNodeConstructor) => {
    return class MediaStreamAudioSourceNode extends noneAudioDestinationNodeConstructor {
        constructor(context, options) {
            const nativeContext = getNativeContext(context);
            const mergedOptions = Object.assign({}, DEFAULT_OPTIONS, options);
            const nativeMediaStreamAudioSourceNode = nativeContext.createMediaStreamSource(mergedOptions.mediaStream);
            super(context, nativeMediaStreamAudioSourceNode, null);
            // Bug #63: Edge & Firefox do not expose the mediaStream yet.
            this._mediaStream = mergedOptions.mediaStream;
            this._nativeMediaStreamAudioSourceNode = nativeMediaStreamAudioSourceNode;
        }
        get mediaStream() {
            return (this._nativeMediaStreamAudioSourceNode.mediaStream === undefined) ?
                this._mediaStream :
                this._nativeMediaStreamAudioSourceNode.mediaStream;
        }
    };
};
